File: runtimes/ios/Neft/Renderer.swift
import UIKit

class Renderer {
    var app: GameViewController!

    var items: [Item] = []

    private let measureTransform = CGAffineTransform(a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0)
    private var dirtyRects = [CGRect]()

    var navigator: Navigator!
    var device: Device!
    var screen: Screen!

    func getItemFromReader(_ reader: Reader) -> Item? {
        let id = reader.getInteger()
        return id == -1 ? nil : self.items[id]
    }

    func load() {
        self.navigator = Navigator()
        self.device = Device()
        self.screen = Screen()

        Device.register()
        Screen.register()
        Navigator.register()
        Item.register()
        Image.register()
        Text.register()
        TextInput.register()
        NativeItem.register()
        Rectangle.register()
        Scrollable.register()
    }

    func pxToDp(_ px: CGFloat) -> CGFloat {
        return px / device.pixelRatio
    }

    func dpToPx(_ dp: CGFloat) -> CGFloat {
        return dp * device.pixelRatio
    }

    func pushItem(_ val: Item) {
        app.client.pushInteger(val.id)
    }
}

File: runtimes/android/app/src/main/java/io/neft/renderer/Renderer.java
package io.neft.renderer;

import java.util.ArrayList;
import java.util.List;

import io.neft.client.Reader;
import lombok.Getter;

public class Renderer {
    final private List<Item> items = new ArrayList<>();
    @Getter final private Device device = new Device();
    final private Screen screen = new Screen();
    final private Navigator navigator = new Navigator();

    public Renderer() {
        items.add(null);
    }

    public float pxToDp(float px) {
        return px / device.getPixelRatio();
    }

    public float dpToPx(float dp) {
        return dp * device.getPixelRatio();
    }

    synchronized int registerItem(Item item) {
        items.add(item);
        return items.size() - 1;
    }

    public Item getItemFromReader(Reader reader) {
        return getItemById(reader.getInteger());
    }

    public Item getItemById(int id) {
        return items.get(id);
    }

    public void init() {
        Device.init(device);
        Screen.init(screen);
        Navigator.init(navigator);
    }
}
